src/Compiler/Microsoft.AspNetCore.Razor.Language/src/TagHelperBinder.cs

```diff
@@ -92,41 +77,57 @@ public TagHelperBinder(string tagHelperPrefix, IReadOnlyList<TagHelperDescriptor
             }
         }
 
-        Dictionary<TagHelperDescriptor, IReadOnlyList<TagMatchingRuleDescriptor>> applicableDescriptorMappings = null;
-        foreach (var descriptor in descriptors)
+        using var _ = DictionaryPool<TagHelperDescriptor, ImmutableArray<TagMatchingRuleDescriptor>>.GetPooledObject(out var applicableDescriptors);
+
+        // First, try any tag helpers with this tag name.
+        if (_registrations.TryGetValue(tagName, out var matchingDescriptors))
         {
-            // We're avoiding descriptor.TagMatchingRules.Where and applicableRules.Any() to avoid
-            // Enumerator allocations on this hot path
-            List<TagMatchingRuleDescriptor> applicableRules = null;
-            foreach (var rule in descriptor.TagMatchingRules)
-            {
-                if (TagHelperMatchingConventions.SatisfiesRule(tagNameWithoutPrefix, parentTagNameWithoutPrefix, attributes, rule))
-                {
-                    applicableRules ??= new List<TagMatchingRuleDescriptor>();
-                    applicableRules.Add(rule);
-                }
-            }
+            FindApplicableDescriptors(matchingDescriptors, tagNameWithoutPrefix, parentTagNameWithoutPrefix, attributes, applicableDescriptors);
+        }
 
-            if (applicableRules != null && applicableRules.Count > 0)
-            {
-                applicableDescriptorMappings ??= new Dictionary<TagHelperDescriptor, IReadOnlyList<TagMatchingRuleDescriptor>>();
-                applicableDescriptorMappings[descriptor] = applicableRules;
-            }
+        // Next, try any "catch all" descriptors.
+        if (_registrations.TryGetValue(TagHelperMatchingConventions.ElementCatchAllName, out var catchAllDescriptors))
+        {
+            FindApplicableDescriptors(catchAllDescriptors, tagNameWithoutPrefix, parentTagNameWithoutPrefix, attributes, applicableDescriptors);
         }
 
-        if (applicableDescriptorMappings == null)
+        if (applicableDescriptors.Count == 0)
         {
             return null;
         }
 
-        var tagHelperBinding = new TagHelperBinding(
+        return new TagHelperBinding(
             tagName,
             attributes,
             parentTagName,
-            applicableDescriptorMappings,
-            _tagHelperPrefix);
+            applicableDescriptors.ToFrozenDictionary(),
+            TagHelperPrefix);
+
+        static void FindApplicableDescriptors(
+            IEnumerable<TagHelperDescriptor> descriptors,
+            ReadOnlySpan<char> tagNameWithoutPrefix,
+            ReadOnlySpan<char> parentTagNameWithoutPrefix,
+            ImmutableArray<KeyValuePair<string, string>> attributes,
+            Dictionary<TagHelperDescriptor, ImmutableArray<TagMatchingRuleDescriptor>> applicableDescriptors)
+        {
+            foreach (var descriptor in descriptors)
+            {
+                using var applicableRules = new PooledArrayBuilder<TagMatchingRuleDescriptor>();
```

Comment:
It seems like we can move this outside the loop?

Reply:
That's true. We could clear the builder on every iteration of the loop.

Reply:
Fixed